src/Sylius/Bundle/ApiBundle/CommandHandler/RegisterShopUserHandler.php

```
@@ -0,0 +1,78 @@
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\CommandHandler;
+
+use Doctrine\Persistence\ObjectManager;
+use Sylius\Bundle\ApiBundle\Command\RegisterShopUser;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Core\Repository\CustomerRepositoryInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\User\Canonicalizer\CanonicalizerInterface;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+final class RegisterShopUserHandler implements MessageHandlerInterface
+{
+    /** @var CanonicalizerInterface */
+    private $canonicalizer;
+
+    /** @var FactoryInterface */
+    private $shopUserFactory;
+
+    /** @var FactoryInterface */
+    private $customerFactory;
+
+    /** @var CustomerRepositoryInterface */
+    private $customerRepository;
+
+    /** @var ObjectManager */
+    private $shopUserManager;
+
+    public function __construct(
+        CanonicalizerInterface $canonicalizer,
+        FactoryInterface $shopUserFactory,
+        FactoryInterface $customerFactory,
+        CustomerRepositoryInterface $customerRepository,
+        ObjectManager $shopUserManager
+    ) {
+        $this->canonicalizer = $canonicalizer;
+        $this->shopUserFactory = $shopUserFactory;
+        $this->customerFactory = $customerFactory;
+        $this->customerRepository = $customerRepository;
+        $this->shopUserManager = $shopUserManager;
+    }
+
+    public function __invoke(RegisterShopUser $command): void
+    {
+        $emailCanonical = $this->canonicalizer->canonicalize($command->email);
+
+        /** @var ShopUserInterface $user */
+        $user = $this->shopUserFactory->createNew();
+        $user->setPlainPassword($command->password);
+
+        /** @var CustomerInterface|null $customer */
+        $customer = $this->customerRepository->findOneBy(['emailCanonical' => $emailCanonical]);
```

If the customer already has a user assigned, we need to throw an exception.